### PR TITLE
Add packager-specific nfpm config (#1829)

### PR DESCRIPTION
**Pull request by jonathanlloyd**
Add new nfpm config options to goreleaser

So that all the options offered by nfpm are available in goreleaser!

Closes: #1829

### CODE REVIEW

**Comment by jonathanlloyd** [comment]
@caarlos0 wyt? Is this what you had in mine in #1829?
If so I'll add apk and rpm

**Comment by caarlos0** [comment]
Overall it looks good, but we need to think how'll we handle the secrets... not sure if we can assume `NFPM_DEB_PASSPHRASE` because the user might have multiple nfpm targets with different signing for some reason (although its a weird use case and sure if its a real one)...

**Comment by jonathanlloyd** [comment]
> Overall it looks good, but we need to think how'll we handle the secrets... not sure if we can assume `NFPM_DEB_PASSPHRASE` because the user might have multiple nfpm targets with different signing for some reason (although its a weird use case and sure if its a real one)...

What about a .goreleaser-secrets.yaml you can put in your .gitignore?
Then you could reference secrets by name:

`.goreleaser-secrets.yaml`
```
deb_passphrase_1: "my-super-secret-passphrase"
deb_passphrase_2: "another-one"
```

`.goreleaser.yaml`
```
nfpms:
 - id: nfpm1
   # other config
   deb:
     key_file: key.gpg
     key_passphrase_secret: deb_passphrase_1
 - id: nfpm2
   # other config
   deb:
     key_file: key.gpg
     key_passphrase_secret: deb_passphrase_2
```

Thoughts?

**Comment by caarlos0** [comment]
I would say to do something similar to what we have on the http pipe: https://goreleaser.com/customization/upload/

Just maybe use `id` instead of `name`... I think that at the time the concept of `id` wasn't implemented yet...

File: www/docs/customization/nfpm.md [hunk 1]
```diff
@@ -156,6 +156,39 @@ nfpms:
           "tmp/app_generated.conf": "/etc/app-rpm.conf"
         scripts:
           preinstall: "scripts/preinstall-rpm.sh"
+
+    # Custon configuration applied only to the Deb packager.
```
Comment (rcurrah) [inline www/docs/customization/nfpm.md:160]: `Custom` not `Custon`

**Comment by caarlos0** [comment]
@jonathanlloyd hey, do you plan on continue working on this? Otherwise I can take it from here :)

**Comment by jonathanlloyd** [comment]
> @jonathanlloyd hey, do you plan on continue working on this? Otherwise I can take it from here :)

Sorry, been super busy. I can take a look at it today :smile:

**Comment by jonathanlloyd** [comment]
@caarlos0 I've updated the env var logic and added some tests
If you're happy I'll do the other packagers :slightly_smiling_face:

**Review by caarlos0 — APPROVED** [review]
So far looks good to me, thanks 🙏🚀

**Comment by vandot** [comment]
After this is merged I will create PR to add deb `templates`, or if my [PR to nfpm](https://github.com/goreleaser/nfpm/pull/240) is merged before we can include it here?

File: internal/pipe/nfpm/nfpm.go [hunk 2]
```diff
@@ -173,6 +174,42 @@ func create(ctx *context.Context, fpm config.NFPM, format, arch string, binaries
 				PreRemove:   overridden.Scripts.PreRemove,
 				PostRemove:  overridden.Scripts.PostRemove,
 			},
+			Deb: nfpm.Deb{
+				Scripts: nfpm.DebScripts{
+					Rules: overridden.Deb.Scripts.Rules,
```
Comment (vandot) [inline internal/pipe/nfpm/nfpm.go:179]: Add
```
Templates: overridden.Deb.Scripts.Templates,
```

File: pkg/config/config.go [hunk 3]
```diff
@@ -323,6 +323,70 @@ type NFPMScripts struct {
...
+	Group       string `yaml:"group,omitempty"`
+	Compression string `yaml:"compression,omitempty"`
+	// https://www.cl.cam.ac.uk/~jw35/docs/rpm_config.html
+	ConfigNoReplaceFiles map[string]string `yaml:"config_noreplace_files,omitempty"`
+	Signature            NFPMRPMSignature  `yaml:"signature,omitempty"`
+}
+
+// NFPMDebScripts is scripts only available on deb packages.
+type NFPMDebScripts struct {
+	Rules string `yaml:"rules,omitempty"`
```
Comment (vandot) [inline pkg/config/config.go:343]: Add
```
Templates string `yaml:"templates,omitempty"`
```

File: www/docs/customization/nfpm.md [hunk 4]
```diff
@@ -156,6 +156,81 @@ nfpms:
...
+        # The id will be transformed to uppercase.
+        # E.g. If your nfpm id is 'default' then the rpm-specific passphrase
+        # should be set as $NFPM_DEFAULT_RPM_PASSPHRASE
+        key_file: key.gpg
+
+    # Custom configuration applied only to the Deb packager.
+    deb:
+      # Custom deb rules script.
+      scripts:
+        rules: foo.sh
```
Comment (vandot) [inline www/docs/customization/nfpm.md:188]: Change this to this
```
  # Custom deb special files.
  scripts:
    # Deb rules script.
    rules: foo.sh
    # Deb templates file, when using debconf.
    templates: templates
```

**Review by vandot — COMMENTED** [review]
My [PR to nfpm](https://github.com/goreleaser/nfpm/pull/240) got merged, if you would like to include `templates` file as well great and thanks :) If not I would create separate PR after this one is merged.
~~On top of all proposed changes you have to update deps `go mod edit -require=github.com/goreleaser/nfpm@v1.9.0`~~

**Comment by ryancurrah** [comment]
Nice templates 👍 . Super stoked for this.

**Comment by caarlos0** [comment]
updating nfpm already: https://github.com/goreleaser/goreleaser/pull/1883

**Comment by jonathanlloyd** [comment]
> My [PR to nfpm](https://github.com/goreleaser/nfpm/pull/240) got merged, if you would like to include `templates` file as well great and thanks :) If not I would create separate PR after this one is merged.
> ~On top of all proposed changes you have to update deps `go mod edit -require=github.com/goreleaser/nfpm@v1.9.0`~

Merged in the latest version of nfpm and added these changes :)

**Comment by caarlos0** [comment]
Great work everyone, many thanks! 🚀

Would be awesome to add some example usages of these advanced options to docs as well (maybe in the how to session), if someone wants to tackle that, please feel free to do so 🙏

**Comment by vandot** [comment]
@caarlos0 I can for `templates` :)

**Comment by ryancurrah** [comment]
When I get to using the new features I can also create a PR for docs.

**Pull request merged**